### ISSUE TITLE
Middleware causes error: Cannot read property 'isDependencyTreeStatic' of undefined

### ISSUE THREAD

**Issue by jimschmit** [MichalLytek/typegraphql-nestjs#23]
Hello, I want to use a Middleware class (that manages if a field is returned or not based on if the customer has an active premium subscription or not).
```
export class PremiumFeature implements MiddlewareInterface {
  constructor() {}

  async use({ context, info }: ResolverData, next: NextFn) {
    return next();
  }
}
```
for certain fields eg:
```
@ObjectType()
@InputType('BusinessInput')
@Entity()
export class Business extends StandardEntity {

  @UseMiddleware(PremiumFeature)
  @Field(field => [Card], { nullable: true })
  @OneToMany(
    () => Card,
    card => card.business,
  )
  public cards: Card[];
}
```
A sample request:
```
query business($id: Int!) {
  business(id: $id) {
    id
    cards{id}
  }
}
```
returns the following response with the error:
```
{
  "errors": [
    {
      "message": "Cannot read property 'isDependencyTreeStatic' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "business",
        "cards"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'isDependencyTreeStatic' of undefined",
            "    at Object.get (/media/jim/Daten/projets/place_to_be/code/api/node_modules/typegraphql-nestjs/dist/typegraphql-options.factory.js:42:42)",
            "    at IOCContainer.getInstance (/media/jim/Daten/projets/place_to_be/code/api/node_modules/type-graphql/dist/utils/container.js:39:26)",
            "    at dispatchHandler (/media/jim/Daten/projets/place_to_be/code/api/node_modules/type-graphql/dist/resolvers/helpers.js:74:65)",
            "    at Object.applyMiddlewares (/media/jim/Daten/projets/place_to_be/code/api/node_modules/type-graphql/dist/resolvers/helpers.js:88:12)",
            "    at /media/jim/Daten/projets/place_to_be/code/api/node_modules/type-graphql/dist/resolvers/create.js:75:26",
            "    at field.resolve (/media/jim/Daten/projets/place_to_be/code/api/node_modules/apollo-server-core/dist/utils/schemaInstrumentation.js:52:26)",
            "    at resolveField (/media/jim/Daten/projets/place_to_be/code/api/node_modules/graphql/execution/execute.js:464:18)",
            "    at executeFields (/media/jim/Daten/projets/place_to_be/code/api/node_modules/graphql/execution/execute.js:292:18)",
            "    at collectAndExecuteSubfields (/media/jim/Daten/projets/place_to_be/code/api/node_modules/graphql/execution/execute.js:748:10)",
            "    at completeObjectValue (/media/jim/Daten/projets/place_to_be/code/api/node_modules/graphql/execution/execute.js:738:10)"
          ]
        }
      }
    }
  ],
  "data": {
    "business": {
      "id": 59,
      "cards": null
    }
  }
}
```

The resolver class is using resolver inheritance, it contains this function:
```
@Authorized()
  @Query(() => Business, { name: 'business' })
  protected async getOne(@Arg('id', () => Int) id: number) {
    const result =  await this.businessService.get(id, {
      relations: [
        'openingHours',
        'tags',
        'tags.parentTag',
        'tags.childTags',
        'headerImages',
        'cards'
      ],
    });
    return result;
  }
```
I'd be happy to know what I can do to resolve this issue. 
Kind regards.

**Comment by MichalLytek** [comment]
Are you using Nest v8?

**Comment by jimschmit** [comment]
No, v7

**Comment by MichalLytek** [comment]
`constructor() {}`

You should use class-based middlewares only when you need to inject something to the class. Otherwise use the standard function-based approach.

I've found the reason of your issue and will aim to fix that soon 😉

**Comment by yusufkandemir** [comment]
@MichalLytek I am facing the same error. From what I understand and tested, that commit only addresses global middleware. So, `@UseMiddleware(ClassBasedMiddleware)` doesn't work, with or without `@Injectable()` on the middleware class.

Adding `providersMetadataMap.set(provider.metatype, provider);` after if blocks, and using `@Injectable()` on the middleware class makes everything work fine.
https://github.com/MichalLytek/typegraphql-nestjs/blob/240ad8ec9afa7b20a273cfc469161a841c1439ae/src/prepare-options.service.ts#L32-L48

Of course, since items have been selectively assigned to the metadata map, I guess there must be a reason to avoid the solution I've tried. I know you probably don't have much time to work on this project, but can you provide some insights about this? I would like to PR this solution if possible. I need to inject something into the middleware, so I can't use middleware functions, I need to use a class. Any alternative solution is also appreciated!

**Comment by MichalLytek** [comment]
The checks are to collect metadata about TypeGraphQL related stuff. So if we need global middleware metadata, we should have another if block with another checks?
I don't know, hard to guess, it's been a while since I worked on this. I never thought it will be used by community when Nest has built-in graphql competitive implementation.

**Comment by yusufkandemir** [comment]
I am not sure about the details. Function-based middleware can be used per method, the docs mention class-based middleware to be able to use injections, but they can only be used as global middleware and not per-method middleware. I am using TypeGraphQL with Nest integration instead of Nest's own implementation because this is the original one, you were planning great features, and there is the Prisma integration.

Anyways, since class-based middleware doesn't work, it's not possible to use Nest's middleware, interceptor, etc. system with TypeGraphQL, and `createMethodDecorator` doesn't expose the `target` to be able to inject some stuff onto it, I was basically left with two options: Migrate away from TypeGraphQL to Nest's own implementation or find other ways to achieve this by creating a plain decorator:

```ts
// See https://stackoverflow.com/a/60608856 for implementation details of a decorator with injections
export function LogWhatever() {
  const injectPrisma = Inject(PrismaService);

  return (
    target: object,
    propertyKey: string,
    propertyDescriptor: PropertyDescriptor,
  ) => {
    injectPrisma(target, 'prisma');

    const originalMethod = propertyDescriptor.value as (
      ...args: any[]
    ) => Promise<Whatever>;

    propertyDescriptor.value = async function (...args: any[]) {
      // @ts-expect-error We can't have accurate types for 'this'
      const prisma = this.prisma as PrismaService;

      // We execute the method, save the return value, use it to create a log entry, and return the result
      const whatever = await originalMethod.apply(this, args);

      void createWhateverLog(prisma, whatever);

      return whatever;
    };
  };
}
```

It doesn't seem to possible to inject the GraphQL context, info, etc. so it seems that the only way out when needing those is to migrate away from TypeGraphQL unfortunately.

**Comment by MichalLytek** [comment]
> Adding providersMetadataMap.set(provider.metatype, provider); after if blocks, and using @Injectable() on the middleware class makes everything work fine.

You can implement that and use your fork safely, until I have some time to fix/implement that properly 😉

**Comment by valerii15298** [comment]
@yusufkandemir I am the same as you using typegraphql and typegraphql-prisma with Nest.js.
Class based middleware really does not work. The workaround I found is to use graphql context instead of dependency injection and use function based middleware.

@MichalLytek Not only my opinion the stack Nest.js + type-graphql + typegraphql-prisma + typegraphql-nestjs is one of the best really. And the only problem in it is the `typegraphql-nestjs` package.

**Comment by MichalLytek** [comment]
Reopening to track non-global middlewares 👀

**Comment by nathan2slime** [comment]
I think I have the same problem, when using class in my authCheck

**Comment by nathan2slime** [comment]
```
export class CustomAuthChecker implements AuthCheckerInterface<ContextType> {
  // inject dependency
  constructor(private readonly userRepository: Repository<User>) {}

  check({ root, args, context, info }: ResolverData<ContextType>, roles: string[]) {
    const userId = getUserIdFromToken(context.token);
    // use injected service
    const user = this.userRepository.getById(userId);

    // custom logic here, e.g.:
    return user % 2 === 0;
  }
}
```
It also doesn't work if I add the inject decorator in the class

**Comment by nathan2slime** [comment]
I still don't know if I should create a new issue about this, but here is my App Module

```
@Module({
  imports: [
    TypeOrmModule.forRoot(config),
    TypeGraphQLModule.forRoot({
      driver: ApolloDriver,
      validate: false,
      emitSchemaFile: {
        path: path.resolve(__dirname, './schema.graphql'),
      },
      authChecker: CustomAuthChecker,
      authMode: 'null',
      context: (ctx: ContextPayloadType) =>
        getAuthContext(ctx.req.headers.authorization),
    }),
    UserModule,
  ],
})
export class AppModule {}
```

**Comment by nathan2slime** [comment]
I think I have the same problem, when using class in my authCheck

```
export class CustomAuthChecker implements AuthCheckerInterface<ContextType> {
  // inject dependency
  constructor(private readonly userRepository: Repository<User>) {}

  check({ root, args, context, info }: ResolverData<ContextType>, roles: string[]) {
    const userId = getUserIdFromToken(context.token);
    // use injected service
    const user = this.userRepository.getById(userId);

    // custom logic here, e.g.:
    return user % 2 === 0;
  }
}
```
It also doesn't work if I add the inject decorator in the class. I still don't know if I should create a new issue about this, but here is my App Module

```
@Module({
  imports: [
    TypeOrmModule.forRoot(config),
    TypeGraphQLModule.forRoot({
      driver: ApolloDriver,
      validate: false,
      emitSchemaFile: {
        path: path.resolve(__dirname, './schema.graphql'),
      },
      authChecker: CustomAuthChecker,
      authMode: 'null',
      context: (ctx: ContextPayloadType) =>
        getAuthContext(ctx.req.headers.authorization),
    }),
    UserModule,
  ],
})
export class AppModule {}
```

> Here is the error returned by the query using the Authorized decorator

```
{
  "errors": [
    {
      "message": "Cannot read properties of undefined (reading 'isDependencyTreeStatic')",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "Login"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "TypeError: Cannot read properties of undefined (reading 'isDependencyTreeStatic')",
          "    at Object.get (D:\\Projects\\goland\\node_modules\\typegraphql-nestjs\\dist\\prepare-options.service.js:47:38)",
          "    at IOCContainer.getInstance (D:\\Projects\\goland\\node_modules\\type-graphql\\dist\\utils\\container.js:39:26)",
          "    at D:\\Projects\\goland\\node_modules\\type-graphql\\dist\\helpers\\auth-middleware.js:9:57",
          "    at dispatchHandler (D:\\Projects\\goland\\node_modules\\type-graphql\\dist\\resolvers\\helpers.js:83:30)",
          "    at applyMiddlewares (D:\\Projects\\goland\\node_modules\\type-graphql\\dist\\resolvers\\helpers.js:89:12)",
          "    at D:\\Projects\\goland\\node_modules\\type-graphql\\dist\\resolvers\\create.js:28:47",
          "    at field.resolve (D:\\Projects\\goland\\node_modules\\@apollo\\server\\src\\utils\\schemaInstrumentation.ts:82:22)",
          "    at executeField (D:\\Projects\\goland\\node_modules\\graphql\\execution\\execute.js:481:20)",
          "    at D:\\Projects\\goland\\node_modules\\graphql\\execution\\execute.js:377:22",
          "    at promiseReduce (D:\\Projects\\goland\\node_modules\\graphql\\jsutils\\promiseReduce.js:23:9)"
        ]
      }
    }
  ],
  "data": null
}
```

For now I'm passing a function to authCheck, but I don't have the dependency injection option 😑

**Comment by valerii15298** [comment]
@nathan2slime you are injecting `userRepository` in the graphql context and then receiving it in your authChecker function right?

This way you only use dependency injection to get dependencies and pass them to context. Recently I moved all my dependencies to the graphql context and do not use DI in the resolvers classes directly, BCS context already contains everything I need.
Tbh, this way it is even slightly more comfortable bcs I do not need to do additional imports and explicitly specify types in the resolver constructor.

**Comment by nathan2slime** [comment]
@valerii15298 I don't understand the last part, specifically the term BCS.
I'm injecting userRepository into my User Module, but since authChecker is a function, I need to create a userRepository instance from a DataSource already created inside the authChecker function.

```
export const customAuthChecker: AuthChecker<ContextDataType> = async (
   { context },
   roles
) => {
   const userRespository = dataSource.getRepository(User);
   const userService = new UserService(userRespository);

  return false;
}
```

But I need the authChecker to be a class, as it says that it is possible in the documentation to take advantage of the userRepository's dependency injection, and not need to create a Data Source to use in the authChecker:

![Screenshot 2023-05-04 085301](https://user-images.githubusercontent.com/105957472/236196163-f0aefa1a-65d2-4e7e-8692-5b0f2f17385e.png)

Here the userRepository injection in my User Module:
```
@Module({
           triangular: [TypeOrmModule.forFeature([User])],
           providers: [UserResolver, UserService],
})
default UserModule export class {}
```
I don't know if I'm doing the injection correctly, but the error is simply caused by the authChecker being a class

**Comment by valerii15298** [comment]
@nathan2slime let me suggest you another workaround: instead of creating instances manually you can use this option https://github.com/MichalLytek/typegraphql-nestjs#forrootasync where you can import any modules and get specific services injected:
```typescript

@Module({
  imports: [
    ConfigModule,
    RecipeModule,
    TypeGraphQLModule.forRootAsync({
      driver: ApolloDriver,
      import: [UserModule], // module to which UserService belongs
      inject: [UserService],
      useFactory: async (userService: UserService) => ({
       context: (_req, _res) => ({userService})
      }),
    }),
  ],
})
export default class AppModule {}
```

And then in your authChecker you can access `context.userService`. You can of course inject any other deps from nestjs which you need TypeGraphQL.

I found this approach even more comfortable for me, since you can inject all your dependencies in one place and you are not tightly coupled with nestjs. Also when you access `context.userService` it is automatically typed with Typescript so no need to specify type for it, hence a bit smaller room for error.

@

**Comment by valerii15298** [comment]
@nathan2slime from you example, maybe try to export it:
```typescript
@Module({
  triangular:[TypeOrmModule.forFeature([User])],
  providers: [UserResolver, UserService],
  exports: [UserService]
})
default UserModule export class {}
```

**Comment by nathan2slime** [comment]
Just fixing it for the imports property, in case anyone comes to see it. It worked here thanks for the help!

```
@Module({
  imports: [TypeOrmModule.forFeature([User])],
  providers: [UserResolver, UserService],
  exports: [UserService],
})
export default class UserModule {}
```

**Comment by nathan2slime** [comment]
Class based non-global middlewares not working here either. Any updates on this issue? I thank

**Comment by valerii15298** [comment]
@nathan2slime thats expected, you can see above in the thread(only global middlewares work with nest DI). 

But for all dependencies you can use approach with graphql context that I pointed out above.

When @MichalLytek will have time he will work on it, but with ability to use graphql context for all deps I don't think it is a priority feature.

**Comment by nathan2slime** [comment]
Got it, thanks a lot then 🙌

**Comment by valerii15298** [comment]
Hello, @MichalLytek 

In process of using typograph-nestjs in our app I noticed that actually every nests dependency that we wanna use in typegraphql middleware(or resolver) we can access via graphql context. For me it even makes more sense. Because I have a single bridge between nestjs and typegrapql, which makes setup really framework agnostic. If I want to switch from nestjs to some other framework I can do it quickly, just by passing deps to context. It kinda like a contract, bridge(see example above).
I do not see a single scenario where I cannot achieve dependency access via extract from context versus DI(btw when I extract dependency from context it automatically typed, where with DI I need to write type for every dependency explicitly).

But my point is, do we really need this feature, and if yes then why?

If not, then isn't it easier just remove it from docs(at least for middlewares) and recommend using context instead(as I pointed out above)? Since it is not working anyway.

I would even prefer using everything from context in resolvers instead of DI, though DI with resolvers works without problems.

**Comment by MichalLytek** [comment]
@valerii15298 
Sure you can use context to inject deps to middleware or resolvers. But it's getting hard if you need that dependency in another layer like service, then you rework the Nest's DI in your context creation function.

It might be also a bit tricky to mimick the rest of Nest's DI like scopes, etc.
So the bug has to be fixed, shouldn't be too hard as I think it's just about collecting options and parsing providers and metadata storage.

**Comment by Ravielson** [comment]
> @MichalLytek I am facing the same error. From what I understand and tested, that commit only addresses global middleware. So, `@UseMiddleware(ClassBasedMiddleware)` doesn't work, with or without `@Injectable()` on the middleware class.
> 
> Adding `providersMetadataMap.set(provider.metatype, provider);` after if blocks, and using `@Injectable()` on the middleware class makes everything work fine.
> 
> https://github.com/MichalLytek/typegraphql-nestjs/blob/240ad8ec9afa7b20a273cfc469161a841c1439ae/src/prepare-options.service.ts#L32-L48
> 
> Of course, since items have been selectively assigned to the metadata map, I guess there must be a reason to avoid the solution I've tried. I know you probably don't have much time to work on this project, but can you provide some insights about this? I would like to PR this solution if possible. I need to inject something into the middleware, so I can't use middleware functions, I need to use a class. Any alternative solution is also appreciated!

**I want to confirm or deny this statement @MichalLytek Can we use class middleware in UseMiddleware, e.g. that implements MiddlewareInterface**

I have such middleware:

`export class RefreshTokenGuardHandMade implements MiddlewareInterface<Context> {
  constructor(
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
    private readonly authService: AuthService,
  ) {}

  async use({ context }: ResolverData<Context>, next: NextFn) {
    const JWT_REFRESH_SECRET = process.env.JWT_REFRESH_SECRET;
    const request = context.req;
    const token =
      request.cookies.refresh_token ||
      request.headers?.authorization?.split(' ')?.[1];
    if (!request || !JWT_REFRESH_SECRET || !token) return false;
    try {
      const payload = await this.jwtService.verifyAsync(token, {
        secret: this.configService.get('JWT_REFRESH_SECRET'),
      });
      if (!payload) {
        return false;
      }
      const user = await this.authService.validateUser(payload.sub);
      console.log('user', user);
      if (!user) {
        throw HttpError('Unauthorized', HttpStatus.UNAUTHORIZED);
      }
      request.user = user;
    } catch (err) {
      throw new Error(err.message);
    }
    return next();
  }
}`

And I would like to do it before the Logout Mutation, here:

`@Resolver()
export class AuthResolver {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly authService: AuthService,
    private readonly userService: UserService,
  ) {}

  @UseMiddleware(RefreshTokenGuardHandMade) // <-------- Here
  @Mutation(() => Boolean)
  async logout(@Ctx() ctx: Context): Promise<boolean> {
    const res = ctx.res;
    const req = ctx.req;
    return req.user?.id && (await this.authService.clearToken(res, req.user?.id));

  }
// ... 
}`

**Comment by MichalLytek** [comment]
@yusufkandemir 
You were right, it was enough to put all providers into the map 😉
Maybe not optimal to store there non-TypeGraphQL values but this at least it will work with future updates and more injectable class-based stuff.

Closed by ee122c5 🔒
